itest/hdkeychain.go

```diff
@@ -0,0 +1,88 @@
+package itest
+
+import (
+	"github.com/btcsuite/btcd/btcutil/hdkeychain"
+	"github.com/btcsuite/btcd/chaincfg"
+	"github.com/lightningnetwork/lnd/keychain"
+)
+
+func deriveChildren(key *hdkeychain.ExtendedKey, path []uint32) (
+	*hdkeychain.ExtendedKey, error) {
+
+	var currentKey = key
+	for idx, pathPart := range path {
+		derivedKey, err := currentKey.DeriveNonStandard(pathPart)
+		if err != nil {
+			return nil, err
+		}
+
+		// There's this special case in lnd's wallet (btcwallet) where
+		// the coin type and account keys are always serialized as a
+		// string and encrypted, which actually fixes the key padding
+		// issue that makes the difference between DeriveNonStandard and
+		// Derive. To replicate lnd's behavior exactly, we need to
+		// serialize and de-serialize the extended key at the coin type
+		// and account level (depth = 2 or depth = 3). This does not
+		// apply to the default account (id = 0) because that is always
+		// derived directly.
+		depth := derivedKey.Depth()
+		keyID := pathPart - hdkeychain.HardenedKeyStart
+		nextID := uint32(0)
+		if depth == 2 && len(path) > 2 {
+			nextID = path[idx+1] - hdkeychain.HardenedKeyStart
+		}
+		if (depth == 2 && nextID != 0) || (depth == 3 && keyID != 0) {
+			currentKey, err = hdkeychain.NewKeyFromString(
+				derivedKey.String(),
+			)
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			currentKey = derivedKey
+		}
+	}
+
+	return currentKey, nil
+}
+
+type hdKeyRing struct {
```

Comment:
why do we need this `hdKeyRing`? By searching `DeriveNextKey` it seems we've already had several implementations already, can we use one of those instead? Also isn't there an RPC service `RPCKeyRing` we can use?

Reply:
Can you elaborate on this, please? I can't find a suitable in-memory implementation of keychain.KeyRing, keychain.ECDHRing, and input.Signer (the types I need in SignCloseTx) which could be created from seed and passphrase.

To create RPCKeyRing, the following objects are needed: keychain.SecretKeyRing, lnwallet.WalletController, and *lncfg.RemoteSigner. I don't know how to make them from seed and passphrase.

I copy-pasted these two types from chantools. If there is a better way to create them, we can consider using that approach in chantools as well.